### PR TITLE
Basic query engine integration

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request adds _really_ basic query engine integration. Currently it outputs an iterator per series. This needs to be optimized to group series together. The query engine and integration need some refactoring to add additional functionality so this PR gives an initial baseline to work from.

## Control flow

The query engine lives in the `influxql` package and only communicates with the server and database through an unexported database interface called `dbi` (I couldn't think of a better name). This interface gives us two benefits:

* It allows us to test the query engine separately from the underlying storage. This is important for isolating bugs into either the engine or storage layer.

* It gives the server more control over locking since the interface is called during well-defined times.

Execution of a `SELECT` statement involves the following steps:

1. Obtain server read lock.
2. Lookup the database specified by the query.
3. Plan the statement - this involves calls to the `dbi` to look up measurements, series & fields.
4. Release the server read lock.
5. Execute query - this executes iterators created during planning through the engine's map-reduce framework.
6. Stream data from the `Executor`.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -1487,21 +1492,29 @@ func (s *Server) ReadSeries(database, retentionPolicy, name string, tags map[str
...
+// Returns a resultset for each statement in the query.
+// Stops on first execution error that occurs.
+func (s *Server) ExecuteQuery(q *influxql.Query, database string, user *User) Results {
+	// Build empty resultsets.
+	results := make(Results, len(q.Statements))
+
+	// Execute each statement.
+	for i, stmt := range q.Statements {
+		var res *Result
+		switch stmt := stmt.(type) {
```
Comment (otoolep) [inline server.go]: This is hilarious. When I orignally coded this I also used "stmt"' but had assumed it would confuse the compiler. But it seems liked it's ok to use "stmt" at these scopes?
.

Comment (levicook) [inline server.go]: Might be a good place specify both len and capacity: `results := make(Results, len(q.Statements), len(q.Statements))` 

Might be premature optimization though too.

File: database.go [hunk 2]
```diff
@@ -737,3 +740,203 @@ func marshalTags(tags map[string]string) []byte {
 	}
 	return []byte(strings.Join(s, "|"))
 }
+
+// dbi is an interface the query engine to communicate with the database during planning.
```
Comment (dgnorton) [inline database.go]: nit: comment should read "...the query engine uses..."

File: database.go [hunk 3]
```diff
@@ -737,3 +740,203 @@ func marshalTags(tags map[string]string) []byte {
...
+		itr.min = min.UnixNano()
+	}
+	if !max.IsZero() {
+		itr.max = max.UnixNano()
+	}
+
+	// Retrieve the policy.
+	// Ignore if there are no groups created on the retention policy.
+	rp := dbi.db.policies[dbi.db.defaultRetentionPolicy]
+	if len(rp.groups) == 0 {
```
Comment (dgnorton) [inline database.go]: I think `rp.groups` should be renamed to `rp.shardGroups`.

File: server.go (hunk 1 above)
Comment (benbjohnson) [inline server.go]: Scoping rules make sense once you get used to them. You can redeclare identifiers inside inner blocks and you can use the outer declaration until the new one is declared.

https://golang.org/ref/spec#Declarations_and_scope

File: database.go [hunk 4]
```diff
@@ -118,22 +121,22 @@ func NewMeasurement(name string) *Measurement {
...
 	// Only 255 fields are allowed. If we go over that then return an error.
-	if len(m.Fields) > math.MaxUint8 {
+	if len(m.Fields)+1 > math.MaxUint8 {
 		return nil, ErrFieldOverflow
 	}
 
 	// Create and append a new field.
 	f := &Field{
-		ID:   uint8(len(m.Fields)),
+		ID:   uint8(len(m.Fields) + 1),
```
Comment (otoolep) [inline database.go]: Ow. :-)

File: database.go [hunk 5]
```diff
@@ -306,6 +299,16 @@ type Series struct {
 	measurement *Measurement
 }
 
+// match returns true if all tags match the series' tags.
+func (s *Series) match(tags map[string]string) bool {
```
Comment (otoolep) [inline database.go]: It seems `s.Tags` could be a superset of tags. Is this OK?

Comment (otoolep) [inline database.go]: From its use below, looks like it's fine.

File: database.go [hunk 6]
```diff
@@ -737,3 +740,203 @@ func marshalTags(tags map[string]string) []byte {
...
+}
+
+// SeriesTagValues returns a slice of tag values for a series.
+func (dbi *dbi) SeriesTagValues(seriesID uint32, keys []string) []string {
+	// Find series by id.
+	s := dbi.db.series[seriesID]
+
+	// Lookup value for each key.
+	values := make([]string, len(keys))
+	for i, keys := range keys {
```
Comment (otoolep) [inline database.go]: Is `keys` made up `keys` or should it be `key`?

File: database.go [hunk 7]
```diff
@@ -737,3 +740,203 @@ func marshalTags(tags map[string]string) []byte {
...
+	// Find field by name.
+	f := m.FieldByName(field)
+	if f == nil {
+		return 0, influxql.Unknown
+	}
+
+	return f.ID, f.Type
+}
+
+// CreateIterator returns an iterator given a series data id, field id, & field data type.
```
Comment (otoolep) [inline database.go]: Should the comment explain what it iterates *over*?

File: database.go [hunk 8]
```diff
@@ -737,3 +740,203 @@ func marshalTags(tags map[string]string) []byte {
...
+		cur := b.Cursor()
+		itr.k, itr.v = cur.Seek(u64tob(uint64(itr.min)))
+		itr.cur = cur
+	}
+
+	return itr
+}
+
+// iterator represents a series data iterator for a shard.
+// It can iterate over all data for a given time range for multiple series in a shard.
```
Comment (otoolep) [inline database.go]: How can iterate over multiple series if there is only 1 `seriesID` in the iterator struct? Is this an iterator for a specific series or not?

File: database.go [hunk 9]
```diff
@@ -737,3 +740,203 @@ func marshalTags(tags map[string]string) []byte {
...
+		// Exit at the end of the cursor.
+		if k == nil {
+			return 0, nil
+		}
+
+		// Extract timestamp & field value.
+		key = int64(btou64(k))
+		value = unmarshalValue(v, i.fieldID)
+
+		// If timestamp is beyond interval time range then push onto lookahead buffer.
```
Comment (otoolep) [inline database.go]: Can you explain why you have this lookahead buffer? What is going on here?

File: influxdb.go [hunk 10]
```diff
@@ -95,6 +99,10 @@ var (
 
 	// ErrSeriesExists is returned when attempting to set the id of a series by database, name and tags that already exists
 	ErrSeriesExists = errors.New("series already exists")
+
+	// ErrNotExecuted is returned when a statement is not executed in a query.
+	// This can occur when a previous statement in the same query has errorred.
```
Comment (otoolep) [inline influxdb.go]: Nit: typo in comment.

File: database.go (hunk 2 above)
Comment (benbjohnson) [inline database.go]: Fixed: 2f2300f

File: database.go (hunk 3 above)
Comment (benbjohnson) [inline database.go]: Fixed: 2f2300f

File: server.go [hunk 11]
```diff
@@ -249,8 +249,9 @@ func (s *Server) setClient(client MessagingClient) error {
 
 	// Start goroutine to read messages from the broker.
 	if client != nil {
-		s.done = make(chan struct{}, 0)
-		go s.processor(client, s.done)
+		done := make(chan struct{}, 0)
```
Comment (otoolep) [inline server.go]: Why this change?

File: server.go [hunk 12]
```diff
@@ -1521,62 +1534,115 @@ func (s *Server) ExecuteQuery(q *influxql.Query, database string, user *User) (i
...
 		case *influxql.DropContinuousQueryStatement:
 			continue
 		case *influxql.ListContinuousQueriesStatement:
 			continue
+		default:
+			panic(fmt.Sprintf("unsupported statement type: %T", stmt))
 		}
+
+		// If an error occurs then stop processing remaining statements.
+		results[i] = res
```
Comment (otoolep) [inline server.go]: Nice.

File: server.go [hunk 13]
```diff
@@ -1521,62 +1534,115 @@ func (s *Server) ExecuteQuery(q *influxql.Query, database string, user *User) (i
...
+		if res.Err != nil {
+			break
+		}
+	}
+
+	// Fill any empty results after error.
+	for i, res := range results {
```
Comment (otoolep) [inline server.go]: Cool too, exactly what I was hoping to see.

File: server.go [hunk 14]
```diff
@@ -1521,62 +1534,115 @@ func (s *Server) ExecuteQuery(q *influxql.Query, database string, user *User) (i
...
+	for row := range ch {
+		res.Rows = append(res.Rows, row)
+	}
+
+	return res
+}
+
+// plans a selection statement under lock.
+func (s *Server) planSelectStatement(stmt *influxql.SelectStatement, database string) (*influxql.Executor, error) {
+	s.mu.Lock()
```
Comment (otoolep) [inline server.go]: What is this lock really about? Is it about protecting access to `s.databases`, since it's not safe to read that map without a lock? It can't be to ensure that db exists during the "select" because the database could always be deleted by some other query coming in while this "select" is executing, right? Or am I missing something?

**Comment by otoolep** [comment]
Looks great @benbjohnson -- I like the changes made to statement execution. Most of my feedback is questions about what is going on in spots.

File: database.go (hunk 4 above)
Comment (benbjohnson) [inline database.go]: It was a clash with the assumption in `influxql` that a zero field id means that it's missing.

File: database.go (hunk 6 above)
Comment (benbjohnson) [inline database.go]: Nope. That was a typo.

File: database.go (hunk 7 above)
Comment (benbjohnson) [inline database.go]: Fixed.

File: database.go (hunk 8 above)
Comment (benbjohnson) [inline database.go]: It's single series right now. I need to refactor the engine to group series together and then I can fix the `influxdb.iterator`.

File: database.go (hunk 9 above)
Comment (benbjohnson) [inline database.go]: The lookahead is necessary when we add an interval. For example, if we're doing a `GROUP BY time(1h)` then we need to read all the points up to the end of the hour. The next call to the iterator at the end of the interval will grab the point outside of the current interval, realize it's gone too far, then push it back on the buffer for the next iteration call (after we move to the next interval).

File: influxdb.go (hunk 10 above)
Comment (benbjohnson) [inline influxdb.go]: Fixed.

File: server.go (hunk 11 above)
Comment (benbjohnson) [inline server.go]: Just me being paranoid. I like to pass in the local variable reference so there's no chance that `s.done` could change in between the instantiation and the passing of it as an arg. It should be used in a lock so that shouldn't happen but, again, I just get paranoid sometimes. :)

File: server.go (hunk 14 above)
Comment (benbjohnson) [inline server.go]: The lock just exists so we can safely grab the metadata. The shards won't be able to get dropped while there is a query in progress but that's because the Bolt `Tx` grabs a read lock.

Comment (otoolep) [inline server.go]: BoltDB saves the day again! :-)
